[MediaStream] Add support for ImageCapture.getPhotoSettings

eric-carlson opened the pull request:
#### e3abeffd3a45dc0db29736fb462089f85c906a00
<pre>
[MediaStream] Add support for ImageCapture.getPhotoSettings
<a href="https://bugs.webkit.org/show_bug.cgi?id=262466">https://bugs.webkit.org/show_bug.cgi?id=262466</a>
rdar://116322614

Reviewed by Youenn Fablet.

Add ImageCapture.getPhotoSettings and implement support in MockRealtimeVideoSource and
AVVideoCaptureSource.

* LayoutTests/fast/mediastream/image-capture-get-photo-settings-expected.txt: Added.
* LayoutTests/fast/mediastream/image-capture-get-photo-settings.html: Added.
* Source/WebCore/DerivedSources.make:

* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::getPhotoSettings):
* Source/WebCore/Modules/mediastream/ImageCapture.h:
* Source/WebCore/Modules/mediastream/ImageCapture.idl:

* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getPhotoSettings const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:

* Source/WebCore/Modules/mediastream/PhotoSettings.idl: Added.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivate::getPhotoSettings):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:

* Source/WebCore/platform/mediastream/PhotoSettings.h: Added.

* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::getPhotoSettings):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
(WebCore::PhotoSettingsOrError::PhotoSettingsOrError):
(WebCore::PhotoSettingsOrError::operator bool const):

* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::toFillLightMode):
(WebCore::AVVideoCaptureSource::getPhotoSettings):

* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::getPhotoSettings):
(WebCore::MockRealtimeVideoSource::settingsDidChange):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::getPhotoSettings):
(WebKit::UserMediaCaptureManagerProxy::getPhotoSettings):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::getPhotoSettings):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::getPhotoSettings):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:

Canonical link: <a href="https://commits.webkit.org/268945@main">https://commits.webkit.org/268945@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e98e8e779ad2c39bfad3fad5154dd4ad94efa5d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21083 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21460 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22150 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22964 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19614 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21323 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24717 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21649 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20860 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21307 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21032 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18284 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23819 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18190 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19112 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25389 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19271 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19314 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23320 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19860 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16879 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19136 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18944 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5058 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23431 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19711 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
```diff
@@ -1460,6 +1460,11 @@ void RealtimeMediaSource::getPhotoCapabilities(PhotoCapabilitiesHandler&& comple
     completion(PhotoCapabilitiesOrError("Not supported"_s));
 }
 
+void RealtimeMediaSource::getPhotoSettings(PhotoSettingsHandler&& completion)
```
This would be a great application for using NativePromise ; much clearer error/rejection path.

eric-carlson replied:
OK

jyavenard commented on Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
```diff
@@ -416,6 +421,25 @@ struct PhotoCapabilitiesOrError {
     String errorMessage;
 };
 
+struct PhotoSettingsOrError {
```
With a NativePromise you wouldn't need to create this object.

Simply create a NativePromise<PhotoSettings, String> and return this.

eric-carlson replied:
This is the pattern we have used so far in media capture. It may make sense to change it, but I am going to follow the pattern for consistency here and we can change them all at once.

jyavenard commented on Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
```diff
@@ -416,6 +421,25 @@ struct PhotoCapabilitiesOrError {
+        : errorMessage(WTFMove(errorMessage))
+    { }
+
+    operator bool() const { return settings.has_value(); }
```
A typedef to Expected<PhotoSettings, String> would be a better fit rather than have a new container like this again.

jyavenard replied:
Also the String may not be moveable, would need to handle isolatedCopy() on it.

eric-carlson replied:
This code path is all main thread only for now.

youennf commented on LayoutTests/fast/mediastream/image-capture-get-photo-settings.html:
```diff
@@ -0,0 +1,48 @@
+        }, `Check getPhotoSettings()`);
+
+
+
```
Could remove some lines

eric-carlson replied:
Fixed

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -328,6 +330,22 @@ void MediaStreamTrack::getPhotoCapabilities(DOMPromiseDeferred<IDLDictionary<Pho
 
+void MediaStreamTrack::getPhotoSettings(DOMPromiseDeferred<IDLDictionary<PhotoSettings>>&& promise) const
+{
+    m_private->getPhotoSettings()->whenSettled(RunLoop::main(), [protectedThis = Ref { *this }, promise = WTFMove(promise)] (RealtimeMediaSource::PhotoSettingsPromise::Result&& result) mutable {
```
Reusing PhotoSettingsPromise names for different types in in RealtimeMediaSource and ImageCapture makes it slightly confusing.
I would suggest using auto&& here if/when we can.

RealtimeMediaSource::PhotoSettingsPromise::Result would ideally be an ExceptionOr<PhotoSettings>.
It would have been nice to use ExceptionOr<PhotoSettings> to be able to simply write: `promise.settle(WTFMove(result));`
But ExceptionOr is not in platform.

eric-carlson replied:
I agree. I renamed the later `PhotoSettingsNativePromise` and switched to `auto` where possible.

youennf commented on Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:
```diff
@@ -111,6 +111,7 @@ class MediaStreamTrackPrivate final
     const RealtimeMediaSourceCapabilities& capabilities() const;
 
     void getPhotoCapabilities(RealtimeMediaSource::PhotoCapabilitiesHandler&&);
+    Ref<RealtimeMediaSource::RealtimeMediaSource::PhotoSettingsPromise> getPhotoSettings();
```
```suggestion
    Ref<RealtimeMediaSource::PhotoSettingsPromise> getPhotoSettings();
```

eric-carlson replied:
Fixed

youennf commented on Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
```diff
@@ -1460,6 +1461,11 @@ void RealtimeMediaSource::getPhotoCapabilities(PhotoCapabilitiesHandler&& comple
 
+auto RealtimeMediaSource::getPhotoSettings() -> Ref<PhotoSettingsPromise>
+{
+    return PhotoSettingsPromise::createAndReject("Not supported"_s);
```
We should return a NotSupportedError here ideally.

eric-carlson replied:
The spec says to return `OperationError` "If the data cannot be gathered for any reason", which we do

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -467,6 +469,43 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
+    if ([device() hasTorch])
+        photoSettings.fillLightMode = { toFillLightMode([device() torchMode]) };
+
+    m_photoSettings = WTFMove(photoSettings);
```
I would tend to write it:
```
if (!m_photoSettings) {
    ...
}
return PhotoSettingsPromise::createAndResolve(*m_photoSettings);
```

eric-carlson replied:
Fixed

youennf commented on Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
```diff
@@ -202,6 +204,27 @@ void MockRealtimeVideoSource::getPhotoCapabilities(PhotoCapabilitiesHandler&& co
+        photoSettings.fillLightMode = { fillLightMode };
+    }
+
+    m_photoSettings = WTFMove(photoSettings);
```
How about a one liner like:
```
// FIXME: Add redEyeReduction info.
m_photoSettings = PhotoSettings { fillLightMode, settings.width(), settings.height(), { });
```

eric-carlson replied:
Changed

youennf commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -5877,12 +5877,19 @@ header: <WebCore/PhotoCapabilities.h>
     std::optional<WebCore::PhotoCapabilities> capabilities;
     String errorMessage;
 };
-#endif
```
Why this removal?

jyavenard replied:
that's likely the cause of the compilation error for tvOS

eric-carlson replied:
Fixed

youennf commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -536,6 +542,18 @@ void UserMediaCaptureManagerProxy::getPhotoCapabilities(RealtimeMediaSourceIdent
 
+void UserMediaCaptureManagerProxy::getPhotoSettings(RealtimeMediaSourceIdentifier sourceID, GetPhotoSettingsCallback&& handler)
+{
+    if (auto* proxy = m_proxies.get(sourceID)) {
```
I would tend to write it:
```
auto* proxy = m_proxies.get(sourceID);
if (!proxy) {
    handler(Unexpected<String>("Device not available"_s));
    return;
}
```

eric-carlson replied:
Fixed

youennf commented on Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
```diff
@@ -117,6 +118,15 @@ void RemoteRealtimeMediaSourceProxy::getPhotoCapabilities(WebCore::RealtimeMedia
 
+Ref<WebCore::RealtimeMediaSource::PhotoSettingsPromise> RemoteRealtimeMediaSourceProxy::getPhotoSettings()
+{
+    return m_connection->sendWithPromisedReply(Messages::UserMediaCaptureManagerProxy::GetPhotoSettings(identifier()))->whenSettled(RunLoop::main(), [](Messages::UserMediaCaptureManagerProxy::GetPhotoSettings::Promise::Result&& result) {
```
This is nice :)

eric-carlson replied:
Thanks to @jyavenard!

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -328,6 +330,22 @@ void MediaStreamTrack::getPhotoCapabilities(DOMPromiseDeferred<IDLDictionary<Pho
+        if (!result) {
+            // https://w3c.github.io/mediacapture-image/#ref-for-dom-imagecapture-getphotosettings②
+            // If the data cannot be gathered for any reason (for example, the MediaStreamTrack being ended
+            // asynchronously), then reject p with a new DOMException whose name is OperationError, and
```
It would be good to add a test case like this one: call getPhotoSettings(), end the track and check the results.
It is not totally clear what we should do here, probably reject would be best but the current patch does not guarantee it in its current form.

youennf replied:
It would also be good to add a test where we create a RTCPeerConnection based track, in which case we will probably get an OperationError with `not supported error` message, which is up to spec I think but a bit strange (it might be somehow better to have NotSupportedError).

eric-carlson replied:
Great suggestion, I will update the test

eric-carlson committed:
[MediaStream] Add support for ImageCapture.getPhotoSettings
https://bugs.webkit.org/show_bug.cgi?id=262466
rdar://116322614

Reviewed by Youenn Fablet.

Add ImageCapture.getPhotoSettings and implement support in MockRealtimeVideoSource and
AVVideoCaptureSource.

* LayoutTests/fast/mediastream/image-capture-get-photo-settings-expected.txt: Added.
* LayoutTests/fast/mediastream/image-capture-get-photo-settings.html: Added.
* Source/WebCore/DerivedSources.make:

* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::getPhotoSettings):
* Source/WebCore/Modules/mediastream/ImageCapture.h:
* Source/WebCore/Modules/mediastream/ImageCapture.idl:

* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getPhotoSettings const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:

* Source/WebCore/Modules/mediastream/PhotoSettings.idl: Added.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivate::getPhotoSettings):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:

* Source/WebCore/platform/mediastream/PhotoSettings.h: Added.

* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::getPhotoSettings):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
(WebCore::PhotoSettingsOrError::PhotoSettingsOrError):
(WebCore::PhotoSettingsOrError::operator bool const):

* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::toFillLightMode):
(WebCore::AVVideoCaptureSource::getPhotoSettings):

* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::getPhotoSettings):
(WebCore::MockRealtimeVideoSource::settingsDidChange):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::getPhotoSettings):
(WebKit::UserMediaCaptureManagerProxy::getPhotoSettings):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::getPhotoSettings):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:

* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::getPhotoSettings):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:

Canonical link: https://commits.webkit.org/268945@main
